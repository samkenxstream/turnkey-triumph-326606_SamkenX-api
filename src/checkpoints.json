[
  112319,
  112418,
  112435,
  112447,
  112480,
  115603,
  115606,
  122651,
  123136,
  132756,
  132759,
  132887,
  133531,
  133532,
  133533,
  133804,
  133808,
  134068,
  134418,
  134427,
  134438,
  134626,
  134627,
  135242,
  135243,
  135331
]
