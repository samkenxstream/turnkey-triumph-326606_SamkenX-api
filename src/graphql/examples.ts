export default `
query {
  votes {
    voter
    proposal
    choice
  }
}
`;
