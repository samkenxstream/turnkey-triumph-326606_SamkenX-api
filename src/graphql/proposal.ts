import mysql from '../mysql';

export default async function query(parent, { id }) {
  const query = 'SELECT * FROM proposals WHERE id = ? LIMIT 1';
  const [proposal] = await mysql.queryAsync(query, [id]);
  return proposal;
}
