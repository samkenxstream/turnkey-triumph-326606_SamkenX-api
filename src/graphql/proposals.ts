import mysql from '../mysql';

export default async function query() {
  return await mysql.queryAsync('SELECT * FROM proposals');
}
