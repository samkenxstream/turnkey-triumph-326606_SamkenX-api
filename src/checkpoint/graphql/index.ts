import { graphqlHTTP } from 'express-graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { toGql, toQuery } from './utils';

export default function get(types) {
  const rootValue = { Query: toQuery(types) };
  const typeDefs = toGql(types);
  const schema = makeExecutableSchema({ typeDefs, resolvers: rootValue });
  return graphqlHTTP({ schema, rootValue });
}
