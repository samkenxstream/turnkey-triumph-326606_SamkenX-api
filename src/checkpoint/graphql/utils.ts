import { introspectionFromSchema } from 'graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import mysql from '../mysql';

async function queryMulti(parent, args, context, info) {
  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map((w) => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(orderBy, orderDirection, skip, first);
  return await mysql.queryAsync(
    `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ? ? LIMIT ?, ?`,
    params
  );
}

async function querySingle(parent, args, context, info) {
  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

export function toSql(typeDefs) {
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  let sql = 'DROP TABLE IF EXISTS checkpoint;';
  sql += '\nCREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));';
  sql += '\nINSERT checkpoint SET number = 0;';
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      sql += `\n\nDROP TABLE IF EXISTS ${clone.name.toLowerCase()}s;`;
      sql += `\nCREATE TABLE ${clone.name.toLowerCase()}s (`;
      let sqlIndexes = ``;
      clone.fields.forEach((field) => {
        const fieldType = field.type.name;
        let sqlType = 'VARCHAR(128)';
        if (fieldType === 'Int') sqlType = 'INT(128)';
        if (fieldType === 'String') sqlType = 'VARCHAR(128)';
        if (fieldType === 'Text') sqlType = 'TEXT';
        sql += `\n  ${field.name} ${sqlType} NOT NULL,`;
        if (fieldType !== 'Text') sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
      });
      sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);`;
    }
  });
  return sql;
}

export function toGql(typeDefs) {
  let where = '';
  let gql = 'type Query {';
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      where += `\n\ninput Where${clone.name} {`;
      gql += `\n  ${clone.name.toLowerCase()}s(`;
      gql += `\n    first: Int`;
      gql += `\n    skip: Int`;
      gql += `\n    orderBy: String`;
      gql += `\n    orderDirection: String`;
      gql += `\n    where: Where${clone.name}`;
      gql += `\n  ): [${clone.name}]`;
      gql += `\n  ${clone.name.toLowerCase()}(id: String): ${clone.name}`;
      clone.fields.forEach((field) => {
        const fieldType = field.type.name;
        if (fieldType !== 'Text') {
          where += `\n  ${field.name}: ${fieldType}`;
          where += `\n  ${field.name}_in: [${fieldType}]`;
          if (fieldType === 'Int') {
            where += `\n  ${field.name}_gt: ${fieldType}`;
            where += `\n  ${field.name}_gte: ${fieldType}`;
            where += `\n  ${field.name}_lt: ${fieldType}`;
            where += `\n  ${field.name}_lte: ${fieldType}`;
          }
        }
      });
      where += `\n}`;
    }
  });
  gql += `\n}\n${where}\n\n${typeDefs}`;
  return gql;
}

export function toQuery(typeDefs) {
  const query = {};
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      query[`${clone.name.toLowerCase()}s`] = queryMulti;
      query[clone.name.toLowerCase()] = querySingle;
    }
  });
  return query;
}
