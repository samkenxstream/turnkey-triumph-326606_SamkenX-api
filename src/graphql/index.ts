import { graphqlHTTP } from 'express-graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import checkpoint from './checkpoint';
import proposals from './proposals';
import proposal from './proposal';
import votes from './votes';
import defaultQuery from './examples';
import typeDefs from './schema';

const rootValue = {
  Query: {
    checkpoint,
    proposals,
    proposal,
    votes
  }
};
const schema = makeExecutableSchema({ typeDefs, resolvers: rootValue });

export default graphqlHTTP({
  schema,
  rootValue,
  graphiql: { defaultQuery }
});
