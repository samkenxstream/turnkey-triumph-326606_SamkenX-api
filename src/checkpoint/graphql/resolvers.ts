import { Provider } from 'starknet';
import { AsyncMySqlPool } from '../mysql';
import { CheckpointsStore, MetadataId } from '../stores/checkpoints';
import { Logger } from '../utils/logger';

/**
 *
 */
export interface ResolverContext {
  log: Logger;
  mysql: AsyncMySqlPool;
  checkpointsStore: CheckpointsStore;
  starknetProvider: Provider;
}

export async function queryMulti(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(skip, first);

  const query = `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ${orderBy} ${orderDirection} LIMIT ?, ?`;
  log.debug({ sql: query, args }, 'executing multi query');

  return await mysql.queryAsync(query, params);
}

export async function querySingle(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  log.debug({ sql: query, args }, 'executing single query');

  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

export async function queryLastIndexedBlock(_parent, _args, ctx: ResolverContext) {
  return ctx.checkpointsStore.getMetadataNumber(MetadataId.LastIndexedBlock);
}

// returns the latest starknet block visible to this node.
export async function queryLatestStarknetBlock(_parent, _args, ctx: ResolverContext) {
  // this might need some rate limiting to avoid being abused
  // because it is quite slow.
  const latestBlock = await ctx.starknetProvider.getBlock();
  return latestBlock.block_number;
}

const MaxCheckpointQuerySize = 1000;
export async function queryCheckpoints(_parent, args, ctx: ResolverContext) {
  const fromBlock = args.fromBlock || 0;
  const contracts = [args.contract];
  const size = args.size || 100;

  if (size > MaxCheckpointQuerySize) {
    throw new Error(`query 'size' exceeds max value. max value is ${MaxCheckpointQuerySize}`);
  }

  const checkpoints = await ctx.checkpointsStore.getNextCheckpointBlocks(
    fromBlock,
    contracts,
    size
  );
  return checkpoints;
}
