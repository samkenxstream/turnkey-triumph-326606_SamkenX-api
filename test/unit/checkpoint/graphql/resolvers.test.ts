import { DeepMockProxy, mock, mockDeep } from 'jest-mock-extended';
import { queryCheckpoints, ResolverContext } from '../../../../src/checkpoint/graphql/resolvers';
import { CheckpointsStore } from '../../../../src/checkpoint/stores/checkpoints';

describe('queryCheckpoints', () => {
  let ctx: DeepMockProxy<ResolverContext> & ResolverContext;

  beforeEach(() => {
    ctx = mockDeep<ResolverContext>();
  });

  it('should work with appropriate defaults', async () => {
    const args = {
      contract: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3'
    };

    await queryCheckpoints({}, args, ctx);

    expect(ctx.checkpointsStore.getNextCheckpointBlocks.mock.calls).toMatchSnapshot();
  });

  it('should work with custom block arguments', async () => {
    const args = {
      contract: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3',
      fromBlock: 5100,
      size: 500
    };

    await queryCheckpoints({}, args, ctx);

    expect(ctx.checkpointsStore.getNextCheckpointBlocks.mock.calls).toMatchSnapshot();
  });

  it('should throw if size exceeds allowed max block', async () => {
    const args = {
      contract: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3',
      size: 500000
    };

    await expect(queryCheckpoints({}, args, ctx)).rejects.toThrowErrorMatchingSnapshot();
  });
});
