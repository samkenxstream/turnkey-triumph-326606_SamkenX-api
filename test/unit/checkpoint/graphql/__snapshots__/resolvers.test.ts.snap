// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryCheckpoints should throw if size exceeds allowed max block 1`] = `"query 'size' exceeds max value. max value is 1000"`;

exports[`queryCheckpoints should work with appropriate defaults 1`] = `
Array [
  Array [
    0,
    Array [
      "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
    ],
    100,
  ],
]
`;

exports[`queryCheckpoints should work with custom block arguments 1`] = `
Array [
  Array [
    5100,
    Array [
      "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
    ],
    500,
  ],
]
`;
